### ISSUE TITLE
not-callable reported when calling a callable returned by non-trivial property

### ISSUE THREAD

**Issue by jendagroovy** [PyCQA/pylint#1730]
### Steps to reproduce

```python
class ClsA(object):
    pass


class ClsB(object):

    def __init__(self):
        self._some_cls = None

    @property
    def some_cls(self):
        if self._some_cls:
            return self._some_cls

        self._some_cls = ClsA
        return ClsA

    def get_object(self):
        return self.some_cls()
```

### Current behavior

```
pylint -E /tmp/x.py
************* Module x
E: 19,15: self.some_cls is not callable (not-callable)
```
### Expected behavior

```
pylint -E /tmp/x.py
```
No output as the property actually returns a callable. Error should only be reported when there is a certainty, that the property can't return a callable.

### pylint --version output
```
pylint 1.7.4,
astroid 1.5.3
Python 2.7.10 (default, Feb  7 2017, 00:08:15)
[GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.34)]
```

**Comment by acatton** [comment]
Here's a real life example, with type annotations:

```python
import bz2
import enum
import gzip
import typing


class Algo(enum.Enum):
    BZIP = 'bzip'
    GZIP = 'gzip'

    @property
    def compress(self) -> typing.Callable[[bytes], bytes]:
        return {
            self.BZIP: bz2.compress,
            self.GZIP: gzip.compress,
        }[self]


def compress(algo: str, data: bytes) -> bytes:
    return Algo(algo).compress(data)
```

```
$ pylint x.py
No config file found, using default configuration
************* Module x
[…]
E: 20,11: Algo(algo).compress is not callable (not-callable)
[…]
```

**Comment by heyakyra** [comment]
Just came across this as well, and managed to find one SO question on it: https://stackoverflow.com/questions/53598374/pylint-giving-not-callable-error-for-object-property-that-is-callable

Also, is it expected that even without using `@property` instance variables are not callable? 

Could this be related to #1493 ?

**Comment by PCManticore** [comment]
@acatton I can no longer reproduce the bug you were having using the latest not released yet version.

@heyakyra Do you have a reproduction case for which you get this error? The reason this wasn't inferrable for the original poster was the fact that `pylint` does not proper support for control flow understanding, which is why it had an issue with `get_object` and `self.some_class` but other than that `pylint` should be able to infer a callable coming from a property. An example would be great.

**Comment by heyakyra** [comment]
Seems to be caused by using a `NamedTuple` in particular in this case. 
```
class Test:
    def __init__(self, field):
        self.Item = NamedTuple("x", [(field, str)])
        self.item: self.Item

    def set_item(self, item):
        self.item = self.Item(item)
```
^ self.Item is not callable

```

```

**Comment by DanielNoord** [comment]
> Seems to be caused by using a `NamedTuple` in particular in this case.
> 
> ```
> class Test:
>     def __init__(self, field):
>         self.Item = NamedTuple("x", [(field, str)])
>         self.item: self.Item
> 
>     def set_item(self, item):
>         self.item = self.Item(item)
> ```
> 
> ^ self.Item is not callable

Should be fixed with #5537.

This leaves only the original post in this issue open. This is more difficult thought as we infer `_some_cls` to potentially be `None`. That requires some form of control flow inference.

**Comment by Pierre-Sassoulas** [comment]
Closing as it's been fixed in 2.15.0-dev0 or I can't reproduce it.

**Comment by jendagroovy** [comment]
Yeah, I can confirm that, actually looks like to have been fixed in 2.13.5 👍
